# Проект по SQL

**Описание проекта**   
Коронавирус застал мир врасплох, изменив привычный порядок вещей. На какое-то время жители городов перестали выходить на улицу, посещать кафе и торговые центры. Зато стало больше времени для книг. Это заметили стартаперы — и бросились создавать приложения для тех, кто любит читать.
Ваша компания решила быть на волне и купила крупный сервис для чтения книг по подписке.   

**Цель исследования**   
Проанализировать базу данных. В ней — информация о книгах, издательствах, авторах, а также пользовательские обзоры книг. Эти данные помогут сформулировать ценностное предложение для нового продукта.

**Описание данных**   

**Таблица books** содержит данные о книгах:
-	book_id — идентификатор книги;
-	author_id — идентификатор автора;
-	title — название книги;
-	num_pages — количество страниц;
-	publication_date — дата публикации книги;
-	publisher_id — идентификатор издателя.   

**Таблица authors** содержит данные об авторах:
-	author_id — идентификатор автора;
-	author — имя автора.   

**Таблица publishers** содержит данные об издательствах:
-	publisher_id — идентификатор издательства;
-	publisher — название издательства;   

**Таблица ratings** содержит данные о пользовательских оценках книг:
-	rating_id — идентификатор оценки;
-	book_id — идентификатор книги;
-	username — имя пользователя, оставившего оценку;
-	rating — оценка книги.   
 
**Таблица reviews** содержит данные о пользовательских обзорах:
-	review_id — идентификатор обзора;
-	book_id — идентификатор книги;
-	username — имя автора обзора;
-	text — текст обзора.


**Схема данных**
<img src=https://pictures.s3.yandex.net/resources/scheme_1589269096.png><br><hr style="border: 10px solid #aecad4;" />

<h1>Содержание<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Подключение-к-базе-данных" data-toc-modified-id="Подключение-к-базе-данных-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Подключение к базе данных</a></span></li><li><span><a href="#Исследуем-таблицы" data-toc-modified-id="Исследуем-таблицы-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Исследуем таблицы</a></span></li><li><span><a href="#Задания" data-toc-modified-id="Задания-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Задания</a></span><ul class="toc-item"><li><span><a href="#Посчитайте,-сколько-книг-вышло-после-1-января-2000-года;" data-toc-modified-id="Посчитайте,-сколько-книг-вышло-после-1-января-2000-года;-3.1"><span class="toc-item-num">3.1&nbsp;&nbsp;</span>Посчитайте, сколько книг вышло после 1 января 2000 года;</a></span></li><li><span><a href="#Для-каждой-книги-посчитайте-количество-обзоров-и-среднюю-оценку;" data-toc-modified-id="Для-каждой-книги-посчитайте-количество-обзоров-и-среднюю-оценку;-3.2"><span class="toc-item-num">3.2&nbsp;&nbsp;</span>Для каждой книги посчитайте количество обзоров и среднюю оценку;</a></span></li><li><span><a href="#Определите-издательство,-которое-выпустило-наибольшее-число-книг-толще-50-страниц-—-так-вы-исключите-из-анализа-брошюры;" data-toc-modified-id="Определите-издательство,-которое-выпустило-наибольшее-число-книг-толще-50-страниц-—-так-вы-исключите-из-анализа-брошюры;-3.3"><span class="toc-item-num">3.3&nbsp;&nbsp;</span>Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры;</a></span></li><li><span><a href="#Определите-автора-с-самой-высокой-средней-оценкой-книг-—-учитывайте-только-книги-с-50-и-более-оценками;" data-toc-modified-id="Определите-автора-с-самой-высокой-средней-оценкой-книг-—-учитывайте-только-книги-с-50-и-более-оценками;-3.4"><span class="toc-item-num">3.4&nbsp;&nbsp;</span>Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками;</a></span></li><li><span><a href="#Посчитайте-среднее-количество-обзоров-от-пользователей,-которые-поставили-больше-48-оценок." data-toc-modified-id="Посчитайте-среднее-количество-обзоров-от-пользователей,-которые-поставили-больше-48-оценок.-3.5"><span class="toc-item-num">3.5&nbsp;&nbsp;</span>Посчитайте среднее количество обзоров от пользователей, которые поставили больше 48 оценок.</a></span></li></ul></li><li><span><a href="#Вывод" data-toc-modified-id="Вывод-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Вывод</a></span></li></ul></div>

## Подключение к базе данных

In [1]:
# импортируем библиотеки
import pandas as pd
from sqlalchemy import text, create_engine

In [2]:
# устанавливаем параметры
db_config = {'user': 'praktikum_student', # имя пользователя
'pwd': 'Sdf4$2;d-d30pp', # пароль
'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
'port': 6432, # порт подключения
'db': 'data-analyst-final-project-db'} # название базы данных
connection_string = 'postgresql://{user}:{pwd}@{host}:{port}/{db}'.format(**db_config)
# сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'})
# чтобы выполнить SQL-запрос, используем Pandas
con=engine.connect()

## Исследуем таблицы

In [3]:
query = '''SELECT * FROM books'''
books = pd.io.sql.read_sql(sql=text(query), con = con)
query = '''SELECT * FROM authors'''
authors = pd.io.sql.read_sql(sql=text(query), con = con)
query = '''SELECT * FROM publishers'''
publishers = pd.io.sql.read_sql(sql=text(query), con = con)
query = '''SELECT * FROM ratings'''
ratings = pd.io.sql.read_sql(sql=text(query), con = con)
query = '''SELECT * FROM reviews'''
reviews = pd.io.sql.read_sql(sql=text(query), con = con)

In [4]:
dataframes = {
    'books': books,
    'authors': authors,
    'publishers': publishers,
    'ratings': ratings,
    'reviews': reviews,
}
for name, df in dataframes.items():
    print(f'Название датафрейма: {name}, наличие дубликатов: {round(df.duplicated().mean()*100)} %.')
    display(df.head())
    display(df.info())  

Название датафрейма: books, наличие дубликатов: 0 %.


book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 6 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   book_id           1000 non-null   int64 
 1   author_id         1000 non-null   int64 
 2   title             1000 non-null   object
 3   num_pages         1000 non-null   int64 
 4   publication_date  1000 non-null   object
 5   publisher_id      1000 non-null   int64 
dtypes: int64(4), object(2)
memory usage: 47.0+ KB


None

Название датафрейма: authors, наличие дубликатов: 0 %.


author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 636 entries, 0 to 635
Data columns (total 2 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   author_id  636 non-null    int64 
 1   author     636 non-null    object
dtypes: int64(1), object(1)
memory usage: 10.1+ KB


None

Название датафрейма: publishers, наличие дубликатов: 0 %.


publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 340 entries, 0 to 339
Data columns (total 2 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   publisher_id  340 non-null    int64 
 1   publisher     340 non-null    object
dtypes: int64(1), object(1)
memory usage: 5.4+ KB


None

Название датафрейма: ratings, наличие дубликатов: 0 %.


rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6456 entries, 0 to 6455
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   rating_id  6456 non-null   int64 
 1   book_id    6456 non-null   int64 
 2   username   6456 non-null   object
 3   rating     6456 non-null   int64 
dtypes: int64(3), object(1)
memory usage: 201.9+ KB


None

Название датафрейма: reviews, наличие дубликатов: 0 %.


review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2793 entries, 0 to 2792
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   review_id  2793 non-null   int64 
 1   book_id    2793 non-null   int64 
 2   username   2793 non-null   object
 3   text       2793 non-null   object
dtypes: int64(2), object(2)
memory usage: 87.4+ KB


None

В нашем распоряжении 5 таблиц с информацией о книгах, издательствах, авторах, а также пользовательские обзоры книг. Просмотрев общую информацию о данных по каждой таблице, проблем мы не обнаружили. Дубликатов и пропущенных значений нет.

## Задания

### Посчитайте, сколько книг вышло после 1 января 2000 года;

In [5]:
query = '''SELECT COUNT(book_id)
           FROM books
           WHERE EXTRACT(YEAR FROM CAST(publication_date AS timestamp)) >= 2000'''
books_count = pd.io.sql.read_sql(sql=text(query), con = con)
books_count

count
0    821

После 1 января 2000 года вышла 821 книга.

### Для каждой книги посчитайте количество обзоров и среднюю оценку;


In [6]:
query = '''SELECT b.book_id, ROUND(AVG(rating), 1) AS rating_avg, COUNT(review_id) AS reviews_cnt
           FROM books b
           LEFT JOIN ratings rt ON b.book_id=rt.book_id
           LEFT JOIN reviews rv ON b.book_id=rv.book_id
           GROUP BY b.book_id
           ORDER BY book_id
        '''
books_ratings_and_reviews = pd.io.sql.read_sql(sql=text(query), con = con)
books_ratings_and_reviews

book_id  rating_avg  reviews_cnt
0          1         3.7            6
1          2         2.5            2
2          3         4.7            9
3          4         4.5            4
4          5         4.0           24
..       ...         ...          ...
995      996         3.7            9
996      997         3.4           15
997      998         3.2           20
998      999         4.5            4
999     1000         3.8           24

[1000 rows x 3 columns]

Составили список книг с информацией о количестве обзоров и средней оценкой для каждой из них.

### Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры;

In [7]:
query = '''SELECT publisher, book_cnt   
           FROM publishers p
           INNER JOIN (SELECT publisher_id, COUNT(book_id) AS book_cnt
                       FROM books
                       WHERE num_pages > 50
                       GROUP BY publisher_id
                       ORDER BY book_cnt DESC
                       LIMIT 1) top_p ON p.publisher_id=top_p.publisher_id'''
publisher = pd.io.sql.read_sql(sql=text(query), con = con)
publisher

publisher  book_cnt
0  Penguin Books        42

Издательство "Penguin Books" выпустило наибольшее число книг толще 50 страниц. Их количество составило 42 экзепляра.

### Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками;

In [8]:
query = '''WITH
           a AS (SELECT book_id, rating
                 FROM ratings
                 WHERE book_id IN (SELECT book_id
                                   FROM (SELECT book_id, COUNT(rating)
                                         FROM ratings
                                         GROUP BY book_id
                                         HAVING COUNT(rating) >= 50) r)),
            c AS (SELECT author_id, AVG(rating) AS rating_avg
                  FROM books b
                  JOIN a ON b.book_id=a.book_id
                  GROUP BY author_id
                  ORDER BY AVG(rating) DESC
                  LIMIT 1)
            SELECT author, ROUND(rating_avg, 1) AS rating_avg
            FROM authors s
            JOIN c ON s.author_id=c.author_id
           '''
top_author= pd.io.sql.read_sql(sql=text(query), con = con)
top_author

author  rating_avg
0  J.K. Rowling/Mary GrandPré         4.3

J.K. Rowling/Mary GrandPré - автор с самой высокой средней оценкой книг (с 50 и более оценками). Средний рейтинг: 4,3.

### Посчитайте среднее количество обзоров от пользователей, которые поставили больше 48 оценок.

In [9]:
query = '''SELECT AVG(review_id_cnt)
           FROM (SELECT username, COUNT(review_id) AS review_id_cnt
                 FROM reviews
                 WHERE username IN (SELECT username
                                    FROM (SELECT username, COUNT(rating_id) 
                                          FROM ratings
                                          GROUP BY username
                                          HAVING COUNT(rating_id) > 48) r)
                                    GROUP BY username) v'''
publisher = pd.io.sql.read_sql(sql=text(query), con = con)
publisher

avg
0  24.0

Cреднее количество обзоров от пользователей, которые поставили больше 48 оценок - 24.

## Вывод

Изучив базу данных, мы выяснили следующее:
- после 1 января 2000 года было выпущено 821 книга,
- количество обзоров и средний рейтинг для каждой книги,
- "Penguin Books" оказалось самым продуктивным издательством, выпустившим наибольшее число книг толще 50 страниц,
- писательница J.K. Rowling/Mary GrandPré имеет самую высокую среднюю оценку книг с 50 и более оценками,
- пользователи, которые оставили более 48 оценок книгам, в среднем пишут 24 обзора.